clients/unshaded/src/main/java/tachyon/client/block/BufferedBlockInStream.java

```diff
@@ -0,0 +1,244 @@
+    mClosed = false;
+    mContext = BlockStoreContext.INSTANCE;
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (mClosed) {
+      return;
+    }
+    mClosed = true;
+  }
+
+  @Override
+  public int read() throws IOException {
+    checkIfClosed();
+    if (remaining() == 0) {
+      close();
+      return -1;
+    }
+    if (remainingInBuffer() == 0) {
+      updateBuffer();
+    }
+    mPos ++;
+    mBufferPos ++;
+    return BufferUtils.byteToInt(mBuffer.get());
+  }
+
+  @Override
+  public int read(byte[] b) throws IOException {
+    return read(b, 0, b.length);
+  }
+
+  @Override
+  public int read(byte[] b, int off, int len) throws IOException {
+    checkIfClosed();
+    Preconditions.checkArgument(b != null, "Read buffer cannot be null");
+    Preconditions.checkArgument(off >= 0 && len >= 0 && len + off <= b.length,
+        String.format("Buffer length (%d), offset(%d), len(%d)", b.length, off, len));
+    if (len == 0) {
+      return 0;
+    } else if (remaining() == 0) { // End of block
+      return -1;
+    }
+
+    int toRead = (int) Math.min(len, remaining());
+    if (remainingInBuffer() > toRead) { // data is fully contained in the buffer
+      mBuffer.get(b, off, toRead);
+      mPos += toRead;
+      mBufferPos += toRead;
+      return toRead;
+    }
+
+    if (toRead > mBuffer.limit() / 2) { // directly read if request is > one-half buffer size
+      mBufferPos = INVALID_BUFFER_POS; // Invalidate the buffer
+      int bytesRead = directRead(b, off, toRead);
+      mPos += bytesRead;
+      incrementBytesReadMetric(bytesRead);
+      return bytesRead;
+    }
+
+    // For a read <= half the buffer size, fill the buffer first, then read from the buffer.
+    updateBuffer();
+    mBuffer.get(b, off, toRead);
+    mPos += toRead;
+    mBufferPos += toRead;
+    return toRead;
+  }
+
+  @Override
+  public long remaining() {
+    return mBlockSize - mPos;
+  }
+
+  @Override
+  public void seek(long pos) throws IOException {
+    checkIfClosed();
+    Preconditions.checkArgument(pos >= 0, "Seek position is negative: " + pos);
+    Preconditions.checkArgument(pos <= mBlockSize, "Seek position is past end of block: "
+        + mBlockSize);
+    mBufferPos = INVALID_BUFFER_POS;
+    mPos = pos;
+  }
+
+  @Override
+  public long skip(long n) throws IOException {
+    checkIfClosed();
+    if (n <= 0) {
+      return 0;
+    }
+
+    long toSkip = Math.min(remaining(), n);
+    mBufferPos = INVALID_BUFFER_POS;
+    mPos += toSkip;
+    return toSkip;
+  }
+
+  /**
+   * Reads from the data source into the buffer. The buffer should be at position 0 and have len
+   * valid bytes available after this method is called. This method should not modify mBufferPos,
+   * mPos, or increment any metrics.
+   *
+   * @param len length of data to fill in the buffer, must always be <= buffer size
+   * @throws IOException
+   */
+  protected abstract void bufferedRead(int len) throws IOException;
```

Comment:
what if the number of successfully read bytes < len after calling `bufferedRead`, throw exception?
Alternatively, shall we return the number of bytes like directRead?